packages/smart-accounts/src/light-account/accounts/base.ts

```diff
@@ -0,0 +1,361 @@
+      if (calls.length === 1) {
+        const call = calls[0];
+        return encodeFunctionData({
+          abi,
+          functionName: "execute",
+          args: [call.to, call.value ?? 0n, call.data ?? "0x"],
+        });
+      }
+
+      const [targets, values, datas] = calls.reduce(
+        (accum, curr) => {
+          accum[0].push(curr.to);
+          accum[1].push(curr.value ?? 0n);
+          accum[2].push(curr.data ?? "0x");
+          return accum;
+        },
+        [[], [], []] as [Address[], bigint[], Hex[]],
+      );
+
+      return encodeFunctionData({
+        abi,
+        functionName: "executeBatch",
+        args: [targets, values, datas],
+      });
+    },
+
+    async getStubSignature() {
+      const signature =
+        "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
+
+      switch (version) {
+        case "v1.0.1":
+        case "v1.0.2":
+        case "v1.1.0":
+          return signature;
+        case "v2.0.0":
+          return concat([SignaturePrefix.EOA, signature]);
+        default:
+          throw new BaseError(`Unknown version ${type} of ${String(version)}`);
+      }
+    },
+
+    async signMessage({ message }) {
+      const { type, data } = await prepareSignature({
+        type: "personal_sign",
+        data: message,
+      });
+
+      const sig =
+        type === "eth_signTypedData_v4"
+          ? await signTypedData(client, {
+              ...data,
+              account: owner,
+            })
+          : await signMessage(client, { account: owner, message });
+
+      return formatSignature(sig);
+    },
+
+    async signTypedData(params) {
+      const { type, data } = await prepareSignature({
+        type: "eth_signTypedData_v4",
+        data: params as TypedDataDefinition,
+      });
+
+      const sig =
+        type === "eth_signTypedData_v4"
+          ? await signTypedData(client, {
+              ...data,
+              account: owner,
+            })
+          : await signMessage(client, { account: owner, message: data });
+
+      return formatSignature(sig);
+    },
+
+    async signUserOperation(parameters) {
+      const { chainId = client.chain.id, ...userOperation } = parameters;
+      const userOpHash = getUserOperationHash({
+        chainId,
+        entryPointAddress: entryPoint.address,
+        entryPointVersion: entryPoint.version,
+        userOperation: {
+          ...userOperation,
+          sender: accountAddress,
+        },
+      });
+
+      const signature = await signMessage(client, {
+        account: owner,
+        message: { raw: userOpHash },
+      });
+
+      return version === "v2.0.0"
+        ? concatHex([SignaturePrefix.EOA, signature])
+        : signature;
+    },
+
+    extend: {
+      source: type,
+      getLightAccountVersion: () => version,
+      encodeUpgradeToAndCall,
+      prepareSignature,
+      formatSignature,
```

Comment:
@adamegyed do you think we actually need to expose these on the account? it seems to me like they are only used internally here. not sure if moldy did this on purpose or AI did it.

Reply:
I think we intended to expose these publicly to give a similar API to the wallet APIs `wallet_prepareSign` and `wallet_formatSign`. They're also used internally on the backend to implement those methods.

I'm not sure if we need them exposed in the smart account instance though, because this should all be handled by how the account implementation forwards requests to the attached signer and how the result is formatted.

I think its fine to keep for now, makes it easier to implement in wallet server without trying to unbundle the `signMessage` / `signTypedData` functions on the backend. But maybe we could reduce the API surface in the future by cutting this.

Reply:
> I think its fine to keep for now, makes it easier to implement in wallet server without trying to unbundle the signMessage / signTypedData functions on the backend. 

ah, good point.